src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -980,24 +977,33 @@ private void FreePinHandles()
             {
                 if (spe != null)
                 {
-                    if (spe._buffer != null && spe._count > 0)
+                    if (spe.Buffer != null && spe.LongCount > 0)
                     {
                         // This element is a buffer.
-                        sendPacketsDescriptor[descriptorIndex].buffer = Marshal.UnsafeAddrOfPinnedArrayElement(spe._buffer, spe._offset);
-                        sendPacketsDescriptor[descriptorIndex].length = (uint)spe._count;
+                        sendPacketsDescriptor[descriptorIndex].buffer = Marshal.UnsafeAddrOfPinnedArrayElement(spe.Buffer, spe.Offset);
+                        sendPacketsDescriptor[descriptorIndex].length = checked((uint)spe.LongCount);
                         sendPacketsDescriptor[descriptorIndex].flags = (Interop.Winsock.TransmitPacketsElementFlags)spe._flags;
                         descriptorIndex++;
                     }
-                    else if (spe._filePath != null)
+                    else if (spe.FilePath != null)
                     {
                         // This element is a file.
                         sendPacketsDescriptor[descriptorIndex].fileHandle = _sendPacketsFileStreams[fileIndex].SafeFileHandle.DangerousGetHandle();
-                        sendPacketsDescriptor[descriptorIndex].fileOffset = spe._offset;
-                        sendPacketsDescriptor[descriptorIndex].length = (uint)spe._count;
+                        sendPacketsDescriptor[descriptorIndex].fileOffset = spe.LongOffset;
+                        sendPacketsDescriptor[descriptorIndex].length = checked((uint)spe.LongCount);
                         sendPacketsDescriptor[descriptorIndex].flags = (Interop.Winsock.TransmitPacketsElementFlags)spe._flags;
                         fileIndex++;
                         descriptorIndex++;
                     }
+                    else if (spe.FileStream != null)
+                    {
+                        // This element is a file.
+                        sendPacketsDescriptor[descriptorIndex].fileHandle = spe.FileStream.SafeFileHandle.DangerousGetHandle();
+                        sendPacketsDescriptor[descriptorIndex].fileOffset = spe.LongOffset;
+                        sendPacketsDescriptor[descriptorIndex].length = checked((uint)spe.LongCount);
```

Comment:
Is this appropriate? e.g. If you request LongCount=5GB this will throw here, right?
Perhaps we need to revisit the API and remove LongCount?

Reply:
32-bit Linux sendfile also has a 32-bit length. This cast is in pal_networking.c:

https://github.com/dotnet/corefx/blob/da62b6a922e353eb5e2a19e717fee7063d3ac035/src/Native/Unix/System.Native/pal_networking.c#L2386

I think we also want to improve this. Perhaps in a separate PR.

Reply:
Now Windows API does not support `long` length only `uint`. See https://github.com/dotnet/corefx/issues/25354#issuecomment-353207074
What is correct way to works with it now? 
For example, I can create workaround by creating many `sendPacketsDescriptor` with different offset and correct length.

Reply:
I'm confused.  Why are we adding long offset/count if the underlying OS APIs do not support this?

Reply:
My be it will work on 64 Unix/Linux and I can create work around code on windows?

Reply:
First of all, *does* it work on Linux?  Seems like there is some confusion here.

Second, when adding new APIs like this, the burden is to show that the new API provides value over existing APIs.  I don't think we've met that burden.

I'm really confused as to what our goal here is.  Is it just to allow FileStream instead of filename to SendPackets?  Is it to enable long-sized file-based packet elements?  Is it both?

Reply:
I think both: long files and file streams.

Reply:
Okay, then the PR needs to support both.